Introduce OpenAMP/RPMsg-Lite on LPC54114

MaureenHelm opened the pull request:
Introduces support for Open Asymmetric Multiprocessing (OpenAMP) using [RPMsg-Lite](https://github.com/codeauroraforum/rpmsg-lite), which is a lightweight implementation of the protocol targeting small memory footprint devices. Support for [eRPC](https://github.com/EmbeddedRPC/erpc) is not yet included due to build system issues with c++, as well as to limit the scope of changes in a single PR.

- Adds support for the slave core (Cortex M0+) on LPC54114
- Adds an IPM shim driver using the mcux mailbox driver
- Imports RPMsg-Lite into ext/
- Adds IPM and RPMsg-Lite samples

There is an outstanding issue when building the master application, such that the slave core binary cannot be found by the assembly files (incbin.S). This issue occurs only with Zephyr SDK toolchain, but not ARM gcc.

Therefore an extra manual step required between running cmake and make of the master application:
```
$ cmake -DBOARD=lpcxpresso54114 ..
$ cp ../../remote/build/zephyr/core1_image.bin .
$ make
```

Fixes #3065

carlescufi commented:
@skordal, @ioannisg FYI

skordal commented on ext/multicore/rpmsg_lite/lib/rpmsg_lite/porting/platform/lpc5411x/rpmsg_platform_zephyr_ipm.c:
```diff
@@ -0,0 +1,265 @@
+{
+    switch (RL_GET_LINK_ID(vq_id))
+    {
+        case 0:
```
Why is only link ID 0 supported here?

stanislav-poboril replied:
LPC54114 has two cores so there is an expectation that most of the applications would use a single link there. I am going to replace 0 with RL_PLATFORM_LPC5411x_M4_M0_LINK_ID. If somebody wants to use multiple instances of RPMsg-Lite, he or she has to define more link IDs and implement it in the switch statement.

carlescufi commented:
I think the external library should be in `ext/lib/ipc/rpmsg_lite`. @MaureenHelm, @nashif thoughts?

carlescufi commented:
@pizi-nordic FYI

carlescufi commented on arch/arm/core/cortex_m/prep_c.c:
```diff
@@ -33,12 +33,18 @@ _GENERIC_SECTION(.vt_pointer_section) void *_vector_table_pointer;
+	_vector_table_pointer = _vector_start;
+#elif defined(CONFIG_XIP) && (CONFIG_FLASH_BASE_ADDRESS != 0) || \
     !defined(CONFIG_XIP) && (CONFIG_SRAM_BASE_ADDRESS != 0)
+#if defined(CONFIG_CPU_CORTEX_M0_HAS_VECTOR_TABLE_REMAP)
```
I'm not sure this is going to work. Vector Table Remap doesn't mean it has VTOR, but rather a mechanism to remap its table. In particular I think the STM32F0 does something different than VTOR for this. @b0661, @simple2017 can you weigh in here?

carlescufi replied:
The other option is that you redefine the weak function `relocate_vector_table` in your `soc.c` like the STM32F0 code has done, in order to have your own particular version that uses VTOR. I think that might be the simplest.

carlescufi replied:
[Here](https://github.com/zephyrproject-rtos/zephyr/commit/2f7688bfd3d3114ff039b5b3d1795a9388d6f355#diff-ee625053bbccbb0d77aba679bdf6a18f) is the relevant STM32F0 commit if you want to take a look.

carlescufi replied:
I think we should merge [this PR](https://github.com/zephyrproject-rtos/zephyr/pull/5988) and you could then simply set `CPU_CORTEX_M_HAS_VTOR` for this IC.

miyatsu replied:
I don't think this gonna work on STM32F0, check [here](https://github.com/zephyrproject-rtos/zephyr/blob/master/arch/arm/soc/st_stm32/stm32f0/soc.c#L39) as @carlescufi mentioned. Its not really a offset register but a remap address 0x2000_0000 to address 0x0000_0000

stanislav-poboril replied:
So I will remove the changes of `prep_c.c` from this PR and add CPU_CORTEX_M_HAS_VTOR for LPC54414's M0+ core instead. Once [the other PR](https://github.com/zephyrproject-rtos/zephyr/pull/5988) is merged.

carlescufi replied:
Yep. The PR is merged now.

carlescufi reviewed (changes requested):
nitpick, this commit: lpc: Add code to include slave core binary for lpc54114 from mcux 2.2.1

Should probably be prefixed with `ext` instead of `lpc`

nashif commented:
> I think the external library should be in ext/lib/ipc/rpmsg_lite. @MaureenHelm, @nashif thoughts?

yeah, or at least ext/lib/multicore/...

stanislav-poboril commented on drivers/ipm/ipm_mcux_mailbox.c:
```diff
@@ -0,0 +1,173 @@
+/*
```
Ok.

stanislav-poboril commented on ext/multicore/rpmsg_lite/lib/rpmsg_lite/porting/platform/lpc5411x/rpmsg_platform.c:
```diff
@@ -0,0 +1,330 @@
+    env_lock_mutex(lock);
+
+    assert(0 <= isr_counter);
+    if (!isr_counter)
```
I agree. However the files come from RPMsg-Lite project version 1.2.0. So it will be fixed in the next release of RPMsg-Lite and then it can get into Zephyr. I would not change the file in Zephyr now in order for the file content to be the same as in referenced version of RPMsg-Lite.

stanislav-poboril commented on ext/multicore/CMakeLists.txt:
```diff
@@ -0,0 +1 @@
+add_subdirectory_ifdef(MULTICORE_RPMSG_LITE rpmsg_lite)
```
Ok. I can move `ext/multicore/rpmsg_lite/` into `ext/lib/ipc/rpmsg_lite`. It will match naming of samples subfolders then.

stanislav-poboril commented:
> nitpick, this commit: lpc: Add code to include slave core binary for lpc54114 from mcux 2.2.1
> Should probably be prefixed with ext instead of lpc

Ok.

galak reviewed (changes requested):
The commit that imports the startup code for AMP doesn’t exactly match the SDK, also we are importing it from SDK 2.3.0 and the reset of the code is from SDK 2.2.1.  So can we sync this with SDK 2.3.0 and either match the SDK or comment about the change of name, etc in commit message/readme.

galak commented on arch/arm/core/cortex_m/reset.S:
```diff
@@ -57,6 +60,10 @@ SECTION_SUBSEC_FUNC(TEXT,_reset_section,__reset)
 SECTION_SUBSEC_FUNC(TEXT,_reset_section,__start)
 
+#if defined(CONFIG_PLATFORM_SPECIFIC_INIT)
+    bl _PlatformInit
```
Why do we need to call _PlatformInit this early?  Can't we just call it from nxp_lpc54114_init().  Also do we really need this implemented in ASM?

stanislav-poboril replied:
The developer of RPMsg-Lite told me it needs to be run while the code executes from address 0x0000000. This comment may give some information:

`
/* Setup the reset handler pointer (PC) and stack pointer value.
             * This is used once the second core runs its startup code.
             * The second core first boots from flash (address 0x00000000)
             * and then detects its identity (Cortex-M0, slave) and checks
             * registers CPBOOT and CPSTACK and use them to continue the
             * boot process.
             * Make sure the startup code for current core (Cortex-M4) is
             * appropriate and shareable with the Cortex-M0 core!
             */ 
`

MarekNovakNXP replied:
This is not related to RPMsg-Lite, but to the architecture of LPC54114.
The secondary core in LPC54114 SOC needs to have a possibility of platform specific code in startup and cannot only use the generic cortex_m/reset.S

@galak Can you develop on why you disagree with this? Do you think other platforms may not need this?

galak commented on drivers/ipm/ipm_mcux.c:
```diff
@@ -0,0 +1,175 @@
+	struct device *dev = arg;
+	struct mcux_mailbox_data *data = dev->driver_data;
+	mailbox_cpu_id_t cpu_id;
+#if defined(__CM4_CMSIS_VERSION)
```
Rather than having multiple ifdef blocks, can we just do something at the top of this file like:

#if defined(__CM4_CMSIS_VERION)
  #define MAILBOX_ID kMAILBOX_CM4
#else
  #define MAILBOX_ID kMAILBOX_CM0
#endif

stanislav-poboril replied:
ok

galak commented on drivers/ipm/Kconfig:
```diff
@@ -22,3 +22,10 @@ config IPM_QUARK_SE_MASTER
 	  Sets up the initial interrupt mask and clears out all
 	  channels. Should be turned on for one CPU only.
 
+config IPM_MCUX
```
Should this be IPM_MCUX_LPC (fixup below as well if we make this LPC specific)

stanislav-poboril replied:
This is for all SoC's which have HAS_MCUX defined: LPC, Kinetis, i.MX. So it is not only for LPC.

galak reviewed (changes requested):
Various code review comments.

MaureenHelm commented:
> I think the external library should be in ext/lib/ipc/rpmsg_lite. @MaureenHelm, @nashif thoughts?

Fine with me

dbkinder commented on boards/arm/lpcxpresso54114_m0/doc/lpcxpresso54114.rst:
```diff
@@ -0,0 +1,168 @@
+Before you start using Zephyr on the LPCXpresso54114, download and run
+`LPCScrypt`_ to update the LPC-Link2 firmware to the latest version. Serial
+communication problems, such as dropping characters, have been observed with
+older versions of the firmware.
```
can you provide a version number here (what will "older versions" mean in 3-6 months)?

stanislav-poboril replied:
The version is in the primary core doc in master branch now. The secondary core board doc has been updated to reference this main doc.

dbkinder commented on boards/arm/lpcxpresso54114_m0/doc/lpcxpresso54114.rst:
```diff
@@ -0,0 +1,168 @@
+You can debug an application in the usual way. Here is an example for the
+:ref:`hello_world` application.
+
+.. zephyr-app-commands::
```
In the commit message you mentioned an "extra manual step required between running cmake and make".  Does that need to be addressed here?

stanislav-poboril replied:
The need for manual step has been eliminated.

dbkinder replied:
Yea.  You might then want to update the commit message that says the manual step is needed.

stanislav-poboril committed:
ext: Add dual core startup code for lpc54114 based on mcux 2.2.1

The new file amp_startup.S is freely based on parts of the file
startup_LPC54114_cm4.S from mcux 2.2.1.
It contains platform specific initialization code for both cores.

Origin: NXP MCUXpresso SDK 2.2.1
URL: mcux.nxp.com
Maintained-by: External

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
lpc: Added support for slave Cortext-M0+ core of lpc54114 soc

Add soc configuration support and dts files for nxp_lpc54xxx_m0.

Adjusted nxp_lpc54xxx soc linker, configuration and dts files
for the presence of slave core.

Added startup hook support in ARM Cortext-M arch to allow
the initialization of slave core from startup and reset.

Origin: Original

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
mcux: Add MCUX IPM driver for lpc, kinetis and iMX socs

Add driver for MCUX mailbox which can be used for lpcxpresso54114
and other lpc, kinetis and iMX socs.

Origin: Original

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
lpc: Add code to include slave core binary for lpc54114 from mcux 2.2.1

Add ext code to be able to include image of slave core in primary
core image.

Origin: NXP MCUXpresso SDK 2.2.1
URL: mcux.nxp.com
Maintained-by: External

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
sample: Add MCUX IPM sample application

Add MCUX IPM sample application. It can be run on lpcxpresso54114
board at the moment.

First, the slave core image has to be built from
samples/subsys/ipc/ipm_mcux/remote/.
Then, buid primary core image from samples/subsys/ipc/ipm_mcux/master/.

Origin: Original

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
ext: Add RPMsg-Lite

Add RPMsg-Lite version 1.2.0 to ext.

Origin: RPMsg-Lite
License: BSD 3-Clause
URL: https://github.com/codeauroraforum/rpmsg-lite
commit: 3ed2d34376a2d1ec04f59e4a89e201d820b5d5e5
Purpose: Introduction of lightweight implementation of the
         Remote Processor Messaging (RPMsg) protocol to Zephyr.
Maintained-by: External

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
sample: Add RPMsg-Lite sample application

Add RPMsg-Lite sample application. It can be run on lpcxpresso54114
board at the moment. The application sends messages from slave
to primary core.

First, the slave core image has to be built from
samples/subsys/ipc/rpmsg_lite/remote/.
Then, buid primary core image from
samples/subsys/ipc/rpmsg_lite/master/.

Origin: Original

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
lpc: Add support of slave core in lpcxpresso54114 board

Add lpcxpresso54114_m0 board dts, yaml, fixup, docs, configuration,
pinmux and header files.

Origin: Original

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril commented:
> The commit that imports the startup code for AMP doesn’t exactly match the SDK...
AMP startup is freely based on parts of SDK code, I have updated the commit message.

stanislav-poboril commented:
Thanks for the comments. I have made some updates internally, @MaureenHelm will update this PR, then you could review again.

MaureenHelm commented:
> @MaureenHelm will update this PR, then you could review again.

Done

skordal commented on samples/subsys/ipc/ipm_mcux/master/CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+enable_language(C ASM)
+
+target_include_directories(app PRIVATE $ENV{ZEPHYR_BASE}/drivers)
+target_include_directories(app PRIVATE $ENV{ZEPHYR_BASE}/samples/subsys/ipc/ipm_mcux/master/src)
```
This include directory is not needed.

skordal commented on samples/subsys/ipc/rpmsg_lite/master/CMakeLists.txt:
```diff
@@ -0,0 +1,36 @@
+
+target_include_directories(app PRIVATE $ENV{ZEPHYR_BASE}/samples/subsys/ipc/rpmsg_lite)
+target_include_directories(app PRIVATE $ENV{ZEPHYR_BASE}/drivers)
+target_include_directories(app PRIVATE $ENV{ZEPHYR_BASE}/samples/subsys/ipc/rpmsg_lite/master/src)
```
This include directory is not needed.

stanislav-poboril commented:
This PR is superseded by #7161, which contains the same content except for RPMsg-Lite, which hasn't been chosen as Zephyr's preferred AMP solution.

So this PR can be cancelled and you can continue reviewing #7161 instead.
